common/src/main/java/com/microsoft/identity/common/internal/ui/webview/challengehandlers/ClientCertAuthChallengeHandler.java

```diff
@@ -138,19 +148,27 @@ public void invoke(@NonNull UsbYubiKeyDevice device) {
                 synchronized (sDeviceLock) {
                     //set device
                     mDevice = device;
+                    //Reset currentDialog to null if necessary.
+                    //In this case, currentDialog would be an ErrorDialog.
+                    if (mCurrentDialog != null) {
+                        mCurrentDialog.onCancelCba();
+                    }
+                    //Make sure instance variables are reset/set properly and requests are canceled.
                     mDevice.setOnClosed(new Runnable() {
                         @Override
                         public void run() {
                             synchronized (sDeviceLock) {
                                 Logger.verbose(TAG, "A YubiKey device was disconnected");
                                 mDevice = null;
+                                //Remove the YKPiv security provider if it was added.
+                                //Note that Security.removeProvider will silently return if YKPiv doesn't exist.
+                                Security.removeProvider("YKPiv");
                                 //Call onCancel on current dialog, if dialog is showing.
                                 synchronized (sSmartcardDialogLock) {
                                     if (mCurrentDialog != null) {
                                         mCurrentDialog.onCancelCba();
-                                        //Reset dialog to null
-                                        mCurrentDialog = null;
-                                        //TODO: maybe show a dialog that tells user they unplugged too early? I'm not sure this is necessary since I'm thinking in a lot of cases, the user wanted to cancel current flow by unplugging.
+                                        //Show dialog informing users that they have unplugged their device too early.
```

Comment:
Wondering if we can do something like this

``` Java


@ThreadSafe
class DialogHolder {

    private Dialog mCurrentDialog = null;

    public synchronized void showErrorDialog(@NonNull final String title, @NonNull final String message){
        showDialog(new SmartcardErrorDialog(title, message));
    }
    
    public synchronized void dismissDialog(){
        showDialog(null);
    }
    
    public synchronized void showDialog(@Nullable final SmartcardDialog dialog){
        if (mCurrentDialog != null) {
            mCurrentDialog.dismiss();
        }
    
        mCurrentDialog = dialog;
    
        if (mCurrentDialog != null){
            mCurrentDialog.show();
        }
    }
}

```

Reply:
I think this is a good idea. Adding methods for showing picker and pin dialogs, as well as methods to help facilitate the onCancelCBA callback.

Reply:
Also noting here (for the future and myself) that calling `dismiss()` on a dialog multiple times doesn't seem to cause any errors.